src/timely-util/src/reclock.rs

```diff
@@ -0,0 +1,773 @@
+            }
+
+            // Accept new data
+            while let Some(event) = puller.pull() {
+                match event.as_mut() {
+                    Event::Progress(changes) => {
+                        source_upper.update_iter(changes.drain(..));
+                    }
+                    Event::Messages(_, data) => {
+                        // Sorting by the linear extension before computing the chain decomposition
+                        // of the batch leads to an optimal chain decomposition for a number of
+                        // concrete Timestamp implementations that we are interested in.
+                        data.sort_unstable_by(|a: &(D1, FromTime, R), b| a.1.cmp(&b.1));
+                        source_batches.push_back(ChainBatch::from_iter(data.drain(..)));
+                    }
+                }
+            }
+
+            // Reclock accepted data
+            let mut output = output.activate();
+            while let Some((frontier, cap)) = replayer.active_binding() {
+                let mut session = output.session(cap);
+                let into_t = cap.time();
+                source_batches.retain_mut(|batch| {
+                    session.give_iterator(
+                        batch
+                            .extract(frontier)
+                            .map(|(d, t, r)| (logic(d, t), into_t.clone(), r)),
+                    );
+                    // Retain non-empty batches
+                    !batch.is_empty()
+                });
+                // If we won't receive any more data for this binding we can go to the next one
+                if PartialOrder::less_equal(&frontier, &source_upper.frontier()) {
+                    replayer.step();
+                } else {
+                    // Otherwise yield and wait for more data
+                    break;
+                }
+            }
+        }
+    });
+
+    (Box::new(pusher), reclocked.as_collection())
+}
+
+/// A batch of differential updates that vary over some partial order. This type maintains the data
+/// as a set of disjoint chains that allows for efficient extraction of batches given a frontier.
+#[derive(Debug)]
+struct ChainBatch<D, T, R> {
+    /// A list of chains (sets of mutually comparable times) sorted by the partial order.
+    chains: Vec<Vec<(D, T, R)>>,
+}
+
+impl<D, T: Timestamp, R> ChainBatch<D, T, R> {
+    /// Extracts all updates with time not greater or equal to any time in `upper`.
+    fn extract<'a>(
+        &'a mut self,
+        upper: AntichainRef<'a, T>,
+    ) -> impl Iterator<Item = (D, T, R)> + 'a {
+        self.chains
+            .iter_mut()
+            .map(move |chain| {
+                // A chain is a set of mutually comparable elements so we can binary search
+                // to find the prefix that is not beyond the frontier and extract it.
+                let idx = chain.partition_point(|(_, time, _)| !upper.less_equal(time));
+                chain.drain(0..idx)
```

Comment:
This will perform work proportional to `chain.len()` fwiw! You might want a `VecDeque` instead?

Reply:
oops! Good catch. Yes